### ISSUE TITLE
Action rollover fails for missing alias during upgrade

### ISSUE THREAD

**Issue by hyadav5** [elastic/curator#1392]
### Elastic Curator: 5.4.1

Ran into an upgrade scenario problem:

Here is the case:

> 1) We have logging deployment where we use elasticsearch for storage and curator for index management. We have 2 versions of our product 0.5 and 0.75
> 2) In 0.5 deployed on customer site. We have indexes a,b,c with 
> action: create_index
> action: alias (associate this with index created above)
> action: rollover (with some max_age setting say 1d)
> 3) Now the problem starts. Now we want to upgrade 0.75 in customer site.
> 4) 0.75 have new index "catalog". During upgrade scenario we just rollover the previous indexes.
> 5) Our rollover is failing complaing that alias for "catalog" is not found.

=========================

Install and Alias action that runs during "INSTALL"

```
      6:
        action: create_index
        description: "Create the metrics index as named, with the specified extra settings."
        options:
          name: "<metrics-{now{yyyy-MM-dd't'HH.mm.ss}}-000001>"
          continue_if_exception: True
          disable_action: False
          extra_settings:
            settings:
              number_of_shards: 3
              number_of_replicas: 1
              refresh_interval: 20s
      .....
      13:
         action: alias
         description: "Associate metrics alias for an index ."
         options:
           name: metrics
           warn_if_no_indices: True
           continue_if_exception: True
           disable_action: False
         remove:
           filters:
           - filtertype: pattern
             kind: prefix
             value: metrics-
         add:
           filters:
           - filtertype: pattern
             kind: prefix
             value: metrics-
           - filtertype: count
             count: 1
             use_age: True
             source: creation_date
             exclude: False
```

=========================

Rollover action that runs during "UPGRADE"
```
      6:
        action: rollover
        description: "Rollover the index associated with alias metrics"
        options:
          disable_action: False
          name: metrics
          new_index: "<metrics-{now{yyyy-MM-dd't'HH.mm.ss}}>"
          conditions:
            max_age: 1s
          extra_settings:
            index.number_of_shards: 3
            index.number_of_replicas: 1
            index.refresh_interval: 20s
      7:
        action: rollover
        description: "Rollover the index associated with alias catalog"
        options:
          disable_action: False
          name: catalog
          new_index: "<catalog-{now{yyyy-MM-dd't'HH.mm.ss}}>"
          conditions:
            max_age: 1s
          extra_settings:
            index.number_of_shards: 3
            index.number_of_replicas: 1
            index.refresh_interval: 20s
```

=========================

Error:
```
2019-04-27 14:54:47,397 INFO      Trying Action ID: 6, "rollover": Rollover the index associated with alias lrsemetrics
2019-04-27 14:54:47,398 INFO      "<lrsemetrics-{now{yyyy-MM-dd't'HH.mm.ss}}>" is using Elasticsearch date math.
2019-04-27 14:54:47,400 INFO      Performing index rollover
2019-04-27 14:54:47,904 INFO      Action ID: 6, "rollover" completed.

2019-04-27 14:54:47,908 INFO      Trying Action ID: 7, "rollover": Rollover the index associated with alias catalog
2019-04-27 14:54:47,908 INFO      "<catalog-{now{yyyy-MM-dd't'HH.mm.ss}}>" is using Elasticsearch date math.
2019-04-27 14:54:47,910 ERROR     alias "catalog" not found.
2019-04-27 14:54:47,910 ERROR     Failed to complete action: rollover.  <class 'ValueError'>: Unable to perform index rollover with alias "catalog". See previous logs for more details.
```

=========================

My questions are:
1) Do we have any setting in **"action: alias"**, like create alias if not exist. Dont error out if alias is already existing.
Or
2) Do we have any setting in **"action: rollover"**, like create alias if not exist before rollover. Dont error out if alias is already existing.
3) Will the above **"action: rollover"** fails if their is no index corresponding to the alias, or it will just create new index and associate alias with it.

In general, how the upgrade is handled in case of new index in new version.

**Comment by untergeek** [comment]
So, if I understand correctly, your `UPGRADE` scenario ends up with an _index_ named `catalog` rather than an _alias_ named `catalog` which points to an index called `"<catalog-{now{yyyy-MM-dd't'HH.mm.ss}}>"` with the date math interpolated? And in that scenario, of course it won't rollover, because _index_ `catalog` is not an alias.

You have a `create_index` action in your `INSTALL` phase, but add the alias later. This is unnecessary:

```
  6:
    action: create_index
    description: "Create the metrics index as named, with the specified extra settings."
    options:
      name: "<metrics-{now{yyyy-MM-dd't'HH.mm.ss}}-000001>"
      continue_if_exception: True
      disable_action: False
      extra_settings:
        settings:
          number_of_shards: 1
          number_of_replicas: 0
          refresh_interval: 20s
        aliases:
          metrics: {}
```
Which I tested for you (`GET metrics-2019-04-29t15.14.57-000001`):
```
{
  "metrics-2019-04-29t15.14.57-000001" : {
    "aliases" : {
      "metrics" : { }
    },
    "mappings" : { },
    "settings" : {
      "index" : {
        "refresh_interval" : "20s",
        "number_of_shards" : "1",
        "provided_name" : "<metrics-{now{yyyy-MM-dd't'HH.mm.ss}}-000001>",
        "creation_date" : "1556550897307",
        "number_of_replicas" : "0",
        "uuid" : "Ll_ZTMyRRtC4iiyC7kebqA",
        "version" : {
          "created" : "7000099"
        }
      }
    }
  }
}
```

Anyway, you have a _new_ index called `catalog` that you are not creating as a named index with an alias suitable for rollover. You cannot expect it to create this implicitly. Elasticsearch does not allow for that without an index template.

So, to answer question 1, 
```
---
actions:
  10:
    action: create_index
    description: create dummy index indexname
    options:
      name: indexname
  11:
    action: alias
    description: Add alias foo to index indexname
    options:
      name: foo
    add:
      filters:
      - filtertype: pattern
        kind: prefix
        value: index
  12:
    action: alias
    description: Add alias foo to index indexname a second time as proof-of-concept
    options:
      name: foo
    add:
      filters:
      - filtertype: pattern
        kind: prefix
        value: index
```
This runs without error. There is no problem running `add` to an existing or non-existing alias.

The answer to question 2 is an emphatic **No.** You must explicitly create a rollover index before you can use the rollover API, whether that is via a dedicated API call, or in an index template.

The answer to question 3 is that a rollover API call will fail under these conditions (there are other fail conditions, but the following are representative of your line of questioning):

1.  The rollover alias does not exist
2. The alias is not rollable because:
    - The associated index cannot be incremented
    - A dedicated index name was not provided in the rollover API call

> In general, how the upgrade is handled in case of new index in new version.

The bottom line for you is that your `UPGRADE` _must_ include a step which creates the new `"<catalog-{now{yyyy-MM-dd't'HH.mm.ss}}>"` index and associates it with an alias for rollover (ostensibly named `catalog`). There is no way around this that I can see.

**Comment by hyadav5** [comment]
Thanks @untergeek....I got all my queries answered. I have added action for "create_index" and "alias" in upgrade scenario itself. My upgrade is passing now. yaahhhh...!!!
